Remove Memory Accounting.

hlinnaka opened the pull request:
This reverts all the memory accounting related changes to aset.c and
elsewhere.

The Memory Accounts were used to collect memory usage information for
EXPLAIN ANALYZE:

```
postgres=# explain analyze select * from tmp_test_prefix order by data , id offset 1000000 - 5;
                                                                       QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=133749.73..133749.84 rows=5 width=18) (actual time=4529.113..4529.130 rows=5 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=111249.84..133749.84 rows=1000000 width=18) (actual time=1588.647..4467.957 rows=1000000 loops=1)
         Merge Key: data, id
         ->  Sort  (cost=111249.84..113749.84 rows=333334 width=18) (actual time=1586.479..1692.267 rows=334042 loops=1)
               Sort Key: data, id
               Sort Method:  quicksort  Memory: 258027kB
               ->  Seq Scan on tmp_test_prefix  (cost=0.00..11592.00 rows=333334 width=18) (actual time=0.107..152.354 rows=334042 loops=1)
 Planning time: 0.974 ms
   (slice0)    Executor memory: 183K bytes.
   (slice1)    Executor memory: 86052K bytes avg x 3 workers, 86052K bytes max (seg0).  Work_mem: 86009K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 4544.867 ms
(13 rows)
```

The "Exeutor memory" fields came from "peak" balance, from the memory
accounts. To replace that, add tracking of peak memory usage to Memory
Contexts. It's much simpler than having a separate "memory account"
dimension, but is sufficient to show comparable "Executor memory" in
EXPLAIN ANALYZE.

With "set explain_memory_verbosity=detail", you got a much more detailed
report, thanks to memory accounts:

```
postgres=# explain analyze select * from tmp_test_prefix order by data , id offset 1000000 - 5;
                                                                                                                         QUERY PLAN

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------
 Limit  (cost=133749.73..133749.84 rows=5 width=18) (actual time=4574.169..4574.180 rows=5 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=111249.84..133749.84 rows=1000000 width=18) (actual time=1673.472..4512.642 rows=1000000 loops=1)
         Merge Key: data, id
         slice 1, seg 0
           Root: Peak/Cur 0/0 bytes. Quota: 0 bytes.
             Top: Peak/Cur 0/0 bytes. Quota: 0 bytes.
               Main: Peak/Cur 4520/4312 bytes. Quota: 0 bytes.
                 Executor: Peak/Cur 100392/100392 bytes. Quota: 0 bytes.
                   X_Motion: Peak/Cur 21576/21576 bytes. Quota: 102400 bytes.
                     X_Sort: Peak/Cur 81896696/11408 bytes. Quota: 130867200 bytes.
                       X_SeqScan: Peak/Cur 58376/31544 bytes. Quota: 102400 bytes.
                 Deserializer: Peak/Cur 1640/1360 bytes. Quota: 0 bytes.
                 Deserializer: Peak/Cur 7320/5248 bytes. Quota: 0 bytes.
             X_Alien: Peak/Cur 0/0 bytes. Quota: 0 bytes.
             MemAcc: Peak/Cur 1808/1808 bytes. Quota: 0 bytes.
             Rollover: Peak/Cur 1397272/1325152 bytes. Quota: 0 bytes.
             SharedHeader: Peak/Cur 5368/5280 bytes. Quota: 0 bytes.
         slice 1, seg 1
           Root: Peak/Cur 0/0 bytes. Quota: 0 bytes.
             Top: Peak/Cur 0/0 bytes. Quota: 0 bytes.
               Main: Peak/Cur 4520/4312 bytes. Quota: 0 bytes.
                 Executor: Peak/Cur 100392/100392 bytes. Quota: 0 bytes.
                   X_Motion: Peak/Cur 21576/21576 bytes. Quota: 102400 bytes.
                     X_Sort: Peak/Cur 81818488/11408 bytes. Quota: 130867200 bytes.
                       X_SeqScan: Peak/Cur 58376/31544 bytes. Quota: 102400 bytes.
                 Deserializer: Peak/Cur 1640/1360 bytes. Quota: 0 bytes.
                 Deserializer: Peak/Cur 7320/5248 bytes. Quota: 0 bytes.
             X_Alien: Peak/Cur 0/0 bytes. Quota: 0 bytes.
             MemAcc: Peak/Cur 1808/1808 bytes. Quota: 0 bytes.
             Rollover: Peak/Cur 1397272/1325152 bytes. Quota: 0 bytes.
             SharedHeader: Peak/Cur 5368/5280 bytes. Quota: 0 bytes.
...
```

That capability is lost. I don't think it's needed, all memory-intensive
executor nodes should use a separate memory context anyway, and should
print its memory usage in the summary EXPLAIN ANALYZE output anyway. But
if it is needed, we could bring back something similar using memory
contexts by creating a separate memory context for each executor node,
instead of using the same ExecutorState context (es_query_context)
everywhere.

Another feature that goes with the bathwater is the "relinguishing" of
memory from one memory-intensive executor node to another. This essentially
reverts commit 1822c8269f.

hlinnaka commented:
Discussion is on-going at https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/4PT0vJSEld0/EPG4p5XWAwAJ. This PR shows concretely what removing memory accounting does.

soumyadeep2007 commented on src/backend/commands/explain_gp.c:
```diff
@@ -983,7 +933,7 @@ cdbexplain_collectStatsFromNode(PlanState *planstate, CdbExplain_SendStatCtx *ct
 	si->workmemwanted = instr->workmemwanted;
 	si->workfileCreated = instr->workfileCreated;
-	si->peakMemBalance = MemoryAccounting_GetAccountPeakBalance(planstate->memoryAccountId);
+	//si->peakMemBalance = MemoryAccounting_GetAccountPeakBalance(planstate->memoryAccountId);
```
Comment to be deleted.

hlinnaka replied:
The whole `peakMemBalance` is now actually dead. I removed it altogether.

soumyadeep2007 commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -171,12 +167,58 @@
+	Size		nextBlockSize;	/* next block size to allocate */
+	Size		allocChunkLimit;	/* effective chunk size limit */
+	AllocBlock	keeper;			/* if not NULL, keep this block over resets */
+
```
Can we add this comment (or some modification of it)?
```
/* 
	 * GPDB: Memory accounting fields
	 * accountingParent: Each MemoryContext will have a designated MemoryContext
	 * that will act as it's account for tracking all memory allocations/de-allocations
	 * performed within the MemoryContext (captured as localBalance, currentBalance
	 * and peakBalance). 
	 * Typically, the account for a MemoryContext will be one of it's ancestors
	 * in the MemoryContext tree. For MemoryContexts that are accounts, their
	 * accountingParent field will be a self-reference.
	 * 
	 * localBalance: This is the memory allocated (in bytes) for this memory context alone
	 * (i.e. it does not consider the memory allocated for any members of the context's
	 * subtree)
	 * 
	 * currentBalance: This field is only applicable to a MemoryContext designated
	 * as an account. It tracks the current bytes allocated in all of the
	 * subtree MemoryContexts it is responsible for.
	 */
```
In case you are sold on changing the names, please update the comment.

hlinnaka replied:
Changed the names per your suggestion, and added that comment with minor edits. Thanks

soumyadeep2007 commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -1096,7 +694,15 @@ AllocSetDelete(MemoryContext context)
 #endif
 
-	AllocSetReleaseAccountingForAllAllocatedChunks(context);
+	if (set->accountingParent == set)
```
Can we add a static inline bool function:
```
static inline bool isMemoryAccount(AllocSet set)
{
	return set->accountingParent == set;
}
```
And call this from **here** and the asserts inside: `AllocSetGetCurrentUsage`, `AllocSetGetPeakUsage` and `AllocSetSetPeakUsage`?

melanieplageman commented:
Okay, still spelunking through the code and trying out new `explain analyze` with `explain_memory_verbosity`. I got a little confused (haven't reasoned about it much, but found it surprising at first)

I did the following:
```
drop table if exists foo;
create table foo(a int, b int);
insert into foo select i%3, i from generate_series(1,1000)i;
analyze foo;
drop table if exists bar; 
create table bar(c int, d int);
insert into bar select i%5, i from generate_series(1,10000)i;
analyze bar;
```

And got the following:
```
postgres=# explain (analyze, verbose) select count(*) from foo, bar where a = c;
                                                                                       QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=27751.53..27751.54 rows=1 width=8) (actual time=1009.864..1009.865 rows=1 loops=1)
   Output: count(*)
   Memory:  0K bytes avg, 0K bytes max.
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=27751.50..27751.53 rows=1 width=8) (actual time=0.652..1009.851 rows=3 loops=1)
         Output: (PARTIAL count(*))
         Memory:  0K bytes avg, 0K bytes max.
         ->  Partial Aggregate  (cost=27751.50..27751.51 rows=1 width=8) (actual time=336.041..336.041 rows=1 loops=1)
               Output: PARTIAL count(*)
               Memory:  0K bytes avg, 0K bytes max (seg0).
               ->  Hash Join  (cost=237.00..22751.50 rows=666667 width=0) (actual time=2.897..588.795 rows=1334000 loops=1)
                     Hash Cond: (foo.a = bar.c)
                     Executor Memory: 8427kB  Segments: 2  Max: 4237kB (segment 0)
                     work_mem: 8427kB  Segments: 2  Max: 4237kB (segment 0)  Workfile: (0 spilling)
                     Memory:  0K bytes avg, 0K bytes max (seg0).
                     Extra Text: (seg0)   Hash chain length 2000.0 avg, 2000 max, using 3 of 524288 buckets.
                     Extra Text: (seg1)   Hash chain length 2000.0 avg, 2000 max, using 2 of 524288 buckets.
                     ->  Seq Scan on public.foo  (cost=0.00..12.00 rows=334 width=4) (actual time=0.028..0.223 rows=667 loops=1)
                           Output: foo.a, foo.b
                           Memory:  0K bytes avg, 0K bytes max (seg0).
                     ->  Hash  (cost=112.00..112.00 rows=3334 width=4) (actual time=4.084..4.084 rows=6000 loops=1)
                           Output: bar.c
                           Memory:  0K bytes avg, 0K bytes max (seg0).
                           ->  Seq Scan on public.bar  (cost=0.00..112.00 rows=3334 width=4) (actual time=0.067..1.964 rows=6000 loops=1)
                                 Output: bar.c
                                 Memory:  0K bytes avg, 0K bytes max (seg0).
 Planning time: 0.520 ms
   (slice0)    Executor memory: 58K bytes.  Vmem reserved: 9216K bytes.
   (slice1)    Executor memory: 2856K bytes avg x 3 workers, 4297K bytes max (seg0).  Vmem reserved: 11947K bytes avg x 3 workers, 13312K bytes max (seg0).  Work_mem: 4237K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Settings: enable_nestloop=off
 Execution time: 1010.454 ms
(32 rows)
```
Why is `Memory` always `0K bytes...`? 

Side note: why doesn't `explain_memory_verbosity` show up with the other GUCs in `Settings` in explain output?

hlinnaka commented:
On 23/11/2019 00:40, Melanie wrote:
> postgres=# explain (analyze, verbose) select count(*) from foo, bar where a = c;
>                                                                                         QUERY PLAN
> ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
>   Finalize Aggregate  (cost=27751.53..27751.54 rows=1 width=8) (actual time=1009.864..1009.865 rows=1 loops=1)
>     Output: count(*)
>     Memory:  0K bytes avg, 0K bytes max.
>     ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=27751.50..27751.53 rows=1 width=8) (actual time=0.652..1009.851 rows=3 loops=1)
>           Output: (PARTIAL count(*))
>           Memory:  0K bytes avg, 0K bytes max.
>           ->  Partial Aggregate  (cost=27751.50..27751.51 rows=1 width=8) (actual time=336.041..336.041 rows=1 loops=1)
>                 Output: PARTIAL count(*)
>                 Memory:  0K bytes avg, 0K bytes max (seg0).
>                 ->  Hash Join  (cost=237.00..22751.50 rows=666667 width=0) (actual time=2.897..588.795 rows=1334000 loops=1)
>                       Hash Cond: (foo.a = bar.c)
>                       Executor Memory: 8427kB  Segments: 2  Max: 4237kB (segment 0)
>                       work_mem: 8427kB  Segments: 2  Max: 4237kB (segment 0)  Workfile: (0 spilling)
>                       Memory:  0K bytes avg, 0K bytes max (seg0).
>                       Extra Text: (seg0)   Hash chain length 2000.0 avg, 2000 max, using 3 of 524288 buckets.
>                       Extra Text: (seg1)   Hash chain length 2000.0 avg, 2000 max, using 2 of 524288 buckets.
>                       ->  Seq Scan on public.foo  (cost=0.00..12.00 rows=334 width=4) (actual time=0.028..0.223 rows=667 loops=1)
>                             Output: foo.a, foo.b
>                             Memory:  0K bytes avg, 0K bytes max (seg0).
>                       ->  Hash  (cost=112.00..112.00 rows=3334 width=4) (actual time=4.084..4.084 rows=6000 loops=1)
>                             Output: bar.c
>                             Memory:  0K bytes avg, 0K bytes max (seg0).
>                             ->  Seq Scan on public.bar  (cost=0.00..112.00 rows=3334 width=4) (actual time=0.067..1.964 rows=6000 loops=1)
>                                   Output: bar.c
>                                   Memory:  0K bytes avg, 0K bytes max (seg0).
>   Planning time: 0.520 ms
>     (slice0)    Executor memory: 58K bytes.  Vmem reserved: 9216K bytes.
>     (slice1)    Executor memory: 2856K bytes avg x 3 workers, 4297K bytes max (seg0).  Vmem reserved: 11947K bytes avg x 3 workers, 13312K bytes max (seg0).  Work_mem: 4237K bytes max.
>   Memory used:  128000kB
>   Optimizer: Postgres query optimizer
>   Settings: enable_nestloop=off
>   Execution time: 1010.454 ms
> (32 rows)
> ```
> Why is `Memory` always `0K bytes...`?

You get those with explain_memory_verbosity='summary'. Memory accounts 
were used to print those values, so they show up as zeros now. This PR 
really should have removed them.

Or if that information is considered too useful to remove, then it needs 
to be reimplemented.

> Side note: why doesn't `explain_memory_verbosity` show up with the other GUCs in `Settings` in explain output?

Only certain GUCs are printed. The ones that affect query planning, 
mostly. The exact logic is in the gp_guc_list_init() function.

- Heikki

hlinnaka commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -290,8 +318,11 @@ static MemoryContextMethods AllocSetMethods = {
-	AllocSetStats,
-	AllocSetReleaseAccountingForAllAllocatedChunks
+	AllocSetStats
+	,AllocSetSetAccountingRoot
```
I actually put the commas to the beginning on purpose, to keep the last "AllocSetStats" line unmodified from upstream. But I guess that was a bit too cute. Added a comment to note that the last entries are GPDB addtitions, and fixed the formatting.

hlinnaka commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -1905,6 +1429,78 @@ AllocSetStats(MemoryContext context, int level, bool print,
 }
 
+static void
+AllocSetSetAccountingRoot(MemoryContext context)
```
Good idea, changed!

hlinnaka commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -171,12 +167,58 @@
+	AllocBlock	keeper;			/* if not NULL, keep this block over resets */
+
+	struct AllocSetContext *accountingParent;
+	Size		localBalance;
```
> Can we adopt `Size mem_allocated; /* track memory allocated for this context */` from upstream here as opposed to `localBalance`?

Hmm. In PostgreSQL master, mem_allocated is in MemoryContextData rather than AllocSetContextData. Yeah, once we get to v13 we should adopt that, because we don't want to duplicate the effort, but until then I feel that trying to backport that would be more trouble. I added a GPDB_13_MERGE_FIXME comment about that though, to remind us to revisit it when we catch up to v13.

hlinnaka commented:
I updated this patch according to @soumyadeep2007's comments (many thanks, very helpful!), and removed the "Memory" lines that showed "0 kB" that @melanieplageman pointed out. I removed the "detail" and "debug" levels from explain_memory_verbosity, as they didn't do anything anymore.

I believe this PR is now ready. Please take a look!

It was never really clear to me what the old memory accounting stuff accomplished, but I believe this gives sufficient level of detail. If someone objects to the removal of functionality here, please speak up now or forever hold your peace!

hlinnaka commented:
I already posted to the gpdb-dev about this, but copying here to note the current status of this patch:

On 14/02/2020 11:32, Heikki Linnakangas wrote:
> 1. Do we need to add a mode that creates a separate memory context for
> each executor mode, to get something similar to current
> explain_memory_verbosity=detail output?

I went ahead and implemented that, to see how it goes. It was pretty 
straightforward, and I can imagine that it could be useful, so I think 
we should include that. See the second commit on the PR now 
(https://github.com/greenplum-db/gpdb/pull/9065). With that, you get 
output like this:

```
postgres=# set explain_memory_verbosity='detail';
SET
postgres=# explain analyze verbose SELECT * FROM explaintest;
 
QUERY PLAN 

--------------------------------------------------------------------------------------------------------------------------------------------------------
  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..3.30 rows=10 
width=4) (actual time=0.727..0.748 rows=10 loops=1)
    Output: id
    Executor Memory: 36kB  Segments: 1  Max: 36kB (segment -1)
    ->  Seq Scan on public.explaintest  (cost=0.00..3.10 rows=4 width=4) 
(actual time=0.047..0.052 rows=5 loops=1)
          Output: id
          Executor Memory: 100kB  Segments: 3  Max: 34kB (segment 0)
  Planning time: 0.666 ms
    (slice0)    Executor memory: 45K bytes.  Vmem reserved: 14336K bytes.
    (slice1)    Executor memory: 40K bytes avg x 3 workers, 40K bytes 
max (seg0).  Vmem reserved: 13312K bytes avg x 3 workers, 13312K bytes 
max (seg0).
  Memory used:  128000kB
  Optimizer: Postgres query optimizer
  Settings: optimizer=off
  Execution time: 1.993 ms
(13 rows)
```
The per-node "Executor Memory" lines are added with with 
explain_memory_verbosity='detail'. Without it, you only get the summary 
lines for each slice.

soumyadeep2007 commented on src/backend/utils/mmgr/mcxt.c:
```diff
@@ -964,11 +831,8 @@ MemoryContextContainsGenericAllocation(MemoryContext context, void *pointer)
+	// FIXME: this is pretty weak. We should get rid of this function.
+	if (header->context == context)
 	{
 		/*
```
The comment needs an update as sharedHeader pointers are no longer a thing.

hlinnaka replied:
Ah, good catch. In fact, this function is now identical to MemoryContextContain(), so we can remove it altogether.

soumyadeep2007 commented on src/backend/utils/mmgr/mcxt.c:
```diff
@@ -1114,11 +978,9 @@ void *
-#endif
 
 	AssertArg(MemoryContextIsValid(context));
+	AssertNotInCriticalSection(context);
```
This is a no-op since we have #if 0ed it (unless you inserted it with an eye for the future when we will solve the 9_4_FIXME)

hlinnaka replied:
Upstream has it here, so let's avoid differing fromupstream. And indeed when we get around to solve that FIXME, then it will become useful.

tvesely commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -309,389 +373,22 @@ static const unsigned char LogTable256[256] =
 	LT16(8), LT16(8), LT16(8), LT16(8), LT16(8), LT16(8), LT16(8), LT16(8)
 };
 
-
```
I used them a couple times, but I wouldn't really miss them if they were gone. I vote we get rid of them.

hlinnaka commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -1084,7 +715,7 @@ AllocSetReset(MemoryContext context)
  */
 static void
-AllocSetDelete(MemoryContext context)
+AllocSetDelete(MemoryContext context, MemoryContext parent)
```
It's needed because MemoryContextDeleteImpl() clears the 'parent' pointer before caling AllocSetDelete().

hlinnaka commented:
And pushed! Thanks for the review!
